README.md
## about

React version of jQuery.slideToggle. JavaScript animation where height is set on every requestAnimationFrame.
The toggle direction can be reversed during the movement.

[![npm version](https://img.shields.io/npm/v/react-slide-toggle.svg?style=flat-square)](https://www.npmjs.com/package/react-slide-toggle)
[![npm downloads](https://img.shields.io/npm/dm/react-slide-toggle.svg?style=flat-square)](https://www.npmjs.com/package/react-slide-toggle)
[![gzip](https://img.shields.io/bundlephobia/minzip/react-slide-toggle.svg)](https://bundlephobia.com/result?p=react-slide-toggle)
[![license](https://img.shields.io/github/license/kunukn/react-slide-toggle.svg)](https://github.com/kunukn/react-slide-toggle/blob/master/LICENSE)

If you are looking for a CSS transition based alternative, then use this instead
https://github.com/kunukn/react-collapse

## Supported React versions

- React version 16.0+

## demo

- https://codepen.io/kunukn/full/wpepGz/ - cdn example
- https://codesandbox.io/s/react-slide-toggle-y3tpn - npm example

## size

- UMD minified size ~7.8Kb (gzipped ~2.5Kb)

## setup / local demo

- git clone or download
- npm install
- npm start

## info

Default easing is cubicInOut. You can reverse the toggle before the movement completes. Ease in-out works best visually when reverse toggling is to be used.

This should be A11Y friendly, you can test the tabindex by tabbing. The collapsed items should be skipped due to usage of display:none (inert functionality)

JS animation is used for best animation control and possibility of adding interpolation or using advanged easing configuration which you can't with CSS alone. This triggers browser reflows on every requestAnimationFrame. If you have a very long page this might not be the best option to use.

## usage example

Look in App component for inspiration. Apply the styling as needed.

### component example, simple - render prop

```js
import SlideToggle from "react-slide-toggle";
// or
// const SlideToggle = require("react-slide-toggle");

// Apply optional padding to .my-collapsible__content-inner
<SlideToggle
  render={({ toggle, setCollapsibleElement }) => (
    <div className="my-collapsible">
      <button className="my-collapsible__toggle" onClick={toggle}>
        toggle
      </button>
      <div className="my-collapsible__content" ref={setCollapsibleElement}>
        <div className="my-collapsible__content-inner">Collapsible content</div>
      </div>
    </div>
  )}
/>;
```

### component example, simple - function as child

```js
import SlideToggle from "react-slide-toggle";

// Apply optional padding to .my-collapsible__content-inner
<SlideToggle>
  {({ toggle, setCollapsibleElement }) => (
    <div className="my-collapsible">
      <button className="my-collapsible__toggle" onClick={toggle}>
        toggle
      </button>
      <div className="my-collapsible__content" ref={setCollapsibleElement}>
        <div className="my-collapsible__content-inner">Collapsible content</div>
      </div>
    </div>
  )}
</SlideToggle>;
```

### toggle state from outside example

```js
import SlideToggle from "react-slide-toggle";

class MyComponent extends React.Component {
  state = { toggleEvent: 0 };

  onToggle = () => {
    this.setState({ toggleEvent: Date.now() });
  };

  render() {
    return (
      <div>
        <button className="toggle" onClick={this.onToggle}>
          Toggle
        </button>
        <SlideToggle toggleEvent={this.state.toggleEvent}>
          {({ setCollapsibleElement }) => (
            <div className="my-collapsible">
              <div
                className="my-collapsible__content"
                ref={setCollapsibleElement}
              >
                <div className="my-collapsible__content-inner">
                  Collapsible content
                </div>
              </div>
            </div>
          )}
        </SlideToggle>
      </div>
    );
  }
}
```

### component usage example with all options

```js
import SlideToggle from "react-slide-toggle";
import BezierEasing from "bezier-easing"; // optional

const bezierEaseInOutQuart = BezierEasing(0.77, 0, 0.175, 1);

<SlideToggle
  duration={280 /* default 300 */}
  easeCollapse={bezierEaseInOutQuart /* default cubicInOut */}
  easeExpand={bezierEaseInOutQuart /* default cubicInOut */}
  collapsed /* default falsy */
  irreversible /* default falsy */
  noDisplayStyle /* default falsy */
  noOverflowHidden /* default falsy */
  bestPerformance /* default falsy */
  whenReversedUseBackwardEase /* default falsy */
  interpolateOnReverse /* default falsy */
  offsetHeight /* default scrollHeight */
  onExpanded={({ hasReversed }) => {
    /* optional event hook */
  }}
  onExpanding={({ range, progress, hasReversed }) => {
    /* optional event hook */
  }}
  onCollapsed={({ hasReversed }) => {
    /* optional event hook */
  }}
  onCollapsing={({ range, progress, hasReversed }) => {
    /* optional event hook */
  }}
  onMount={({ toggleState, toggle }) => {
    /* optional event hook */
  }}
  onUnmount={({ toggleState }) => {
    /* optional event hook */
  }}
  expandEvent /* default undefined */
  collapseEvent /* default undefined */
  toggleEvent /* default undefined */
  render={({
    toggle,
    setCollapsibleElement,
    toggleState,
    isMoving,
    hasReversed,
    range /* linear value between [0 and 1] */,
    progress /* easing result value between [0 and 1] */
  }) => {
    /* optional logic here */

    /*
      markup example
      where setCollapsibleElement, toggle and progress are used
    */
    return (
      <div className="slide-toggle">
        <div className="slide-toggle__header">
          <button className="slide-toggle__button" onClick={toggle}>
            toggle
          </button>
        </div>
        <div className="slide-toggle__box" ref={setCollapsibleElement}>
          <div
            className="slide-toggle__box-inner"
            style={{ opacity: Math.max(0.5, progress) }}
          >
            Collapsible content
          </div>
        </div>
      </div>
    );
  }}
/>;
```

## properties

- duration - movement duration in milli seconds
- easeCollapse - function which generates a value between [0 and 1]
- easeExpand - function which generates a value between [0 and 1]
- collapsed - start in collapsed mode
- irreversible - you can't reverse direction during movement
- noDisplayStyle - skip adding display:none on collapsed
- noOverflowHidden - skip adding overflow:hidden on the collapsible element
- bestPerformance - don't apply setState for every frame update. Disables range and progress update
- whenReversedUseBackwardEase - play backwards on reverse toggling
- interpolateOnReverse - avoid jumpy height changes when easeCollapse and easeExpand gives far different height position on reverse toggling.
- onExpanded - event hook
- onExpanding - event hook
- onCollapsed - event hook
- onCollapsing - event hook
- onMount - event hook
- onUnmount - event hook
- render - render callback
- children - render callback
- offsetHeight - use offsetHeight HTML element calculation
- expandEvent - update value to `Date.now()` to invoke a controlled expand
- collapseEvent - update value to `Date.now()` to invoke a controlled collapse
- toggleEvent - update value to `Date.now()` to invoke a controlled toggle

## cdn

https://unpkg.com/react-slide-toggle/

```html
<script src="https://unpkg.com/react-slide-toggle/dist/ReactSlideToggle.umd.js">
  <script>
  <script>
  var SlideToggle = window.ReactSlideToggle;
</script>
```

## npm

https://www.npmjs.com/package/react-slide-toggle

## provide your own markup

The component provides the functionality.
Minimum requirement is to bind the collapsible element with `setCollapsibleElement`.
Use the `toggle` function to toggle the collapsible element.

## provide your own easing functions

Look for examples in the App component

```js
import eases from "eases";
import BezierEasing from "bezier-easing";
```

To minimize the component size, no default easing library has been added.

You can see examples of JS-easing library usage here

- eases https://codepen.io/kunukn/full/mpVJOm/
- BezierEasing https://codepen.io/kunukn/full/YYNqyj/

## design goals

- flexible - provide your own markup, styling and easing
- interruptible - can be reversed during movement
- simple api with event hooks
- inert - when collapsed you should tab over the collapsed component
- minimal in size
- availability - from cdn or npm install
- generate range and progress [0;1] values which can be used for further custom animation
- JS is used over CSS transition on purpose to enable possible interpolation or other custom math calculations

## library implementation details

- Used life-cycle are `componentWillUnmount` and `render`
- Extends `React.Component`
- Uses `setState`

## local development

- git clone or download
- npm install
- npm run build
- The build files are now in the dist folder


CHANGELOG.md
## Changelog

### 0.3.5

- bundle contains
  - main: umd build
  - cjs: commonjs build
  - module: esm ES2015 build

### 0.2.5

- add ES2015 bundle

### 0.2.4

- update readme

### 0.2.2

- smaller UMD bundle size
- more bundler options

### 0.2.1

- correct Readme, import path

### 0.2.0

- simplify import
- change props scrollHeight to offsetHeight

### 0.1.13

- add props scrollHeight

### 0.1.12

- add react and react-dom as npm peerDependencies

### 0.1.11

- apply guard clause around usage of collapsible element styles

### 0.1.10

- update readme

### 0.1.9

- fix misspelling and minor refactorings

### 0.1.8

- fix minor bug with noDisplayStyle usage

### 0.1.7

- update demo

### 0.1.6

- add npmignore file

### 0.1.5

- remove console.warn
- don't error silently if wrongly setup
- use umd as default main in dist bundle

### 0.1.4

- add props disableWarnings
- add ability to use function as child

### 0.1.3

- fix irreversible not working

### 0.1.2

- use offsetHeight instead of clientHeight

### 0.1.1

- update readme

### 0.1.0

- update readme

### 0.0.32

- add bestPerformance

### 0.0.31

- add render prop: progress, the easing result value

### 0.0.30

- update readme

### 0.0.29

- fix typo: setCollapsibleElement
- detect window or global object

### 0.0.28

- adjust interpolation

### 0.0.27

- update readme

### 0.0.26

- add interpolateOnReverse

### 0.0.25

- add hasReversed to on event hooks

### 0.0.24

- update deploy files

### 0.0.23

- add render prop: noDisplayStyle

### 0.0.22

- remove boxHeight and updateBoxHeight from render prop

### 0.0.21

- fix range in state

### 0.0.20

- update readme

### 0.0.19

- add render prop: range

### 0.0.18

- cleanup dependencies

### 0.0.17

- use eslint

### 0.0.16

- add render prop: whenReversedUseBackwardEase

### 0.0.15

- changed render prop to hasReversed

### 0.0.14

- add boxHeight and updateBoxHeight to render props

### 0.0.13

- update readme

### 0.0.10

- code refactorings

### 0.0.8

- add callback on event change

### 0.0.7

- ability to reverse toggle direction before the animation is complete.

### 0.0.6

- remove dependency to eases

### 0.0.5

- add deploy script

### 0.0.4

- update readme

### 0.0.3

- make cdn work

### 0.0.2

- demo version added

### 0.0.1

- first version


src/demo/App.js
import React from 'react';
import ReactDOM from 'react-dom';

import eases from 'eases'; // example, provide your own easing fn
import BezierEasing from 'bezier-easing'; // example, provide your own easing fn

const log = console.log.bind(console);

const easeNames = Object.keys(eases);

const easeInOutQuart = t =>
  t < 0.5 ? 8 * t * t * t * t : 1 - 8 * --t * t * t * t;

const bezierEaseInOutQuart = BezierEasing(0.77, 0, 0.175, 1);
const defaultEase = eases.cubicInOut;

const getRandomEase = () => {
  const index = Math.floor(Math.random() * easeNames.length);
  return eases[easeNames[index]];
};

const ToggleText = () => 'Toggle';

const ProgressBar = ({ progress }) => (
  <span className="progress-bar">
    <span
      className="progress-bar__inner"
      style={{
        transform: `scaleX(${progress})`,
      }}
    />
  </span>
);

export default class App extends React.Component {
  state = { duration: 1000, event: 0 };

  generateMarkup = ({
    easeCollapseName,
    easeExpandName,
    easeCollapse,
    easeExpand,
    className = '',
    customText,
  } = {}) => ({
    onToggle,
    setCollapsibleElement,
    toggleState,
    isMoving,
    hasReversed,
    range,
    progress,
  }) => (
    <div
      className={`slide-toggle ${className} ${(
        toggleState || ''
      ).toLowerCase()}`}
    >
      <div className="slide-toggle__header">
        <button className="slide-toggle__toggle" onClick={onToggle}>
          <ToggleText />
        </button>
        <ProgressBar progress={progress} />
      </div>
      <div className="slide-toggle__box" ref={setCollapsibleElement}>
        <div
          className="slide-toggle__box-inner"
          style={{ opacity: Math.max(0.5, range) }}
        >
          <p>
            Default easing is cubicInOut. You can reverse the toggle before the
            movement completes. Ease in-out works best when reverse toggling is
            to be used.
          </p>
          <p>
            This should be A11Y friendly, you can test the tabindex by tabbing.
            The collapsed items should be skipped due to usage of display:none
          </p>
          <p>
            JS animation is used for best animation control and possible easing
            configuration. This triggers browser reflows on every
            requestAnimationFrame. If you have a very long page this might not
            be the best option to use.
          </p>
          <img
            className="slide-toggle__image"
            alt="random"
            src="https://source.unsplash.com/user/erondu/600x200"
          />
          <button onClick={onToggle}>
            <ToggleText />
          </button>
        </div>
      </div>

      <div className="state-values">
        <div>
          <span>toggleState:</span>
          <span>{toggleState}</span>
        </div>
        <div>
          <span>range:</span>
          <span>{range.toFixed(2)}</span>
        </div>
        <div>
          <span>progress:</span>
          <span>{progress.toFixed(2)}</span>
        </div>
        <div>
          <span>isMoving:</span>
          <span>{`${isMoving}`}</span>
        </div>
        <div>
          <span>easeCollapse:</span>
          <span>{easeCollapseName}</span>
        </div>
        <div>
          <span>easeExpand:</span>
          <span>{easeExpandName}</span>
        </div>
        <div>
          <span>hasReversed:</span>
          <span>{`${hasReversed}`}</span>
        </div>
        {customText && (
          <div>
            <span>custom:</span>
            <span>{customText}</span>
          </div>
        )}
      </div>
    </div>
  );

  render() {
    const components = [];
    let SlideToggle = this.props.SlideToggle;
    let SlideToggle2 = this.props.SlideToggle2
      ? this.props.SlideToggle2
      : SlideToggle;
    let SlideToggle3 = this.props.SlideToggle3
      ? this.props.SlideToggle3
      : SlideToggle;
    let SlideToggle4 = this.props.SlideToggle4
      ? this.props.SlideToggle4
      : SlideToggle;

    1 &&
      components.push(
        <div key={components.length}>
          <button onClick={() => this.setState({ event: Date.now() })}>
            Toggle event
          </button>
          <SlideToggle
            key={components.length}
            duration={this.state.duration}
            toggleEvent={this.state.event}
            collapsed
            onMount={({ toggle }) => {
              log('onMount');
              toggle();
            }}
          >
            {this.generateMarkup({
              easeCollapseName: 'default',
              easeExpandName: 'default',
            })}
          </SlideToggle>
        </div>
      );

    1 &&
      components.push(
        <SlideToggle
          key={components.length}
          duration={this.state.duration}
          collapsed
          onCollapsed={() => log('onCollapsed')}
          onCollapsing={() => log('onCollapsing')}
          onExpanding={() => log('onExpanding')}
          onExpanded={() => log('onExpanded')}
        >
          {this.generateMarkup({
            easeCollapseName: 'default',
            easeExpandName: 'default',
          })}
        </SlideToggle>
      );

    1 &&
      components.push(
        <SlideToggle2
          key={components.length}
          duration={this.state.duration * 2}
          easeCollapse={eases.bounceOut}
          easeExpand={eases.bounceOut}
          collapsed
          whenReversedUseBackwardEase
          render={this.generateMarkup({
            easeCollapseName: 'bounceOut',
            easeExpandName: 'bounceOut',
            easeCollapse: eases.bounceOut,
            easeExpand: eases.bounceOut,
            customText: 'whenReversedUseBackwardEase',
          })}
        />
      );

    1 &&
      components.push(
        <SlideToggle3
          key={components.length}
          duration={this.state.duration * 2}
          easeCollapse={eases.expoOut}
          easeExpand={eases.expoOut}
          collapsed
          interpolateOnReverse
          render={this.generateMarkup({
            easeCollapseName: 'expoOut interpolate',
            easeExpandName: 'expoOut interpolate',
            easeCollapse: eases.expoOut,
            easeExpand: eases.expoOut,
            customText: 'interpolateOnReverse',
          })}
        />
      );

    1 &&
      components.push(
        <SlideToggle4
          key={components.length}
          noDisplayStyle
          duration={this.state.duration}
          easeCollapse={bezierEaseInOutQuart}
          easeExpand={bezierEaseInOutQuart}
          collapsed={Math.random() > 0.5}
          render={this.generateMarkup({
            easeCollapseName: this.fnName(bezierEaseInOutQuart),
            easeExpandName: this.fnName(bezierEaseInOutQuart),
            easeCollapse: bezierEaseInOutQuart,
            easeExpand: bezierEaseInOutQuart,
            className: '-header-height',
            customText: 'noDisplayStyle',
          })}
        />
      );

    if (1)
      for (let i = 0; i < 4; i++) {
        const ease = getRandomEase();
        const name = this.fnName(ease);
        components.push(
          <SlideToggle
            key={components.length}
            collapsed
            whenReversedUseBackwardEase
            duration={this.state.duration}
            easeCollapse={ease}
            easeExpand={ease}
            render={this.generateMarkup({
              easeCollapseName: name,
              easeExpandName: name,
              customText: 'whenReversedUseBackwardEase',
            })}
          />
        );
      }

    1 &&
      components.push(
        <SlideToggle
          key={components.length}
          duration={this.state.duration}
          interpolateOnReverse
          easeCollapse={eases.quartOut}
          easeExpand={eases.bounceOut}
          onExpanded={({ hasReversed }) => {
            if (!hasReversed) {
              const el = ReactDOM.findDOMNode(this).querySelector(
                '.-extra-anim .slide-toggle__box-inner'
              );
              if (el) {
                const anim = document.createElement('div');
                const anim2 = document.createElement('div');
                const anim3 = document.createElement('div');
                const anim4 = document.createElement('div');
                const anim5 = document.createElement('div');
                const bcr = el.getBoundingClientRect();
                anim.className = 'bounce-anim';
                anim2.className = 'bounce-anim';
                anim3.className = 'bounce-anim';
                anim4.className = 'bounce-anim';
                anim5.className = 'bounce-anim';
                anim.style.left = `${Math.random() * bcr.width}px`;
                anim2.style.left = `${Math.random() * bcr.width}px`;
                anim3.style.left = `${Math.random() * bcr.width}px`;
                anim4.style.left = `${Math.random() * bcr.width}px`;
                anim5.style.left = `${Math.random() * bcr.width}px`;
                el.appendChild(anim);
                el.appendChild(anim2);
                el.appendChild(anim3);
                el.appendChild(anim4);
                el.appendChild(anim5);
                setTimeout(() => {
                  /* cleanup */
                  anim && el.removeChild(anim);
                  anim2 && el.removeChild(anim2);
                  anim3 && el.removeChild(anim3);
                  anim4 && el.removeChild(anim4);
                  anim5 && el.removeChild(anim5);
                }, 1000);
              }
            }
          }}
          collapsed={Math.random() > 0.5}
          render={this.generateMarkup({
            className: '-extra-anim',
            easeCollapseName: 'quartOut',
            easeExpandName: 'bounceOut',
            easeCollapse: eases.quartOut,
            easeExpand: eases.bounceOut,
            customText: 'interpolateOnReverse',
          })}
        />
      );

    0 &&
      components.push(
        <SlideToggle
          key={components.length}
          duration={this.state.duration}
          collapsed
          scrollHeight
          render={({ onToggle, setCollapsibleElement }) => (
            <div className="my-collapsible">
              <button className="button" onClick={onToggle}>
                toggle
              </button>
              <div
                className="my-collapsible__content"
                ref={setCollapsibleElement}
              >
                Collapsible content
              </div>
            </div>
          )}
        />
      );

    return (
      <>
        <div className="app">
          <header className="app__header">
            <button
              className="app__button"
              onClick={() => {
                this.setState({ duration: ~~(Math.random() * 800 + 200) });
              }}
            >
              Randomize duration
            </button>
            <span>{this.state.duration}</span>
          </header>
          <div className="ease-names">
            <div>Ease names ({easeNames.length}):</div>
            {easeNames.map((name, index) => (
              <span key={index}>{name} </span>
            ))}
          </div>
          {components}
        </div>
      </>
    );
  }

  getFunctionName(fn) {
    return /function ([^(]*)/.exec(`${fn}`)[1];
  }

  fnName = fn => this.getFunctionName(fn);
}
